OpenRA.Mods.Common/Scripting/Properties/UpgradeProperties.cs

```diff
@@ -17,35 +18,41 @@ namespace OpenRA.Mods.Common.Scripting
 	[ScriptPropertyGroup("General")]
 	public class UpgradeProperties : ScriptActorProperties, Requires<UpgradeManagerInfo>
 	{
-		UpgradeManager um;
+		readonly UpgradeManager um;
+		readonly ScriptUpgradesCache validUpgrades;
+
 		public UpgradeProperties(ScriptContext context, Actor self)
 			: base(context, self)
 		{
 			um = self.Trait<UpgradeManager>();
+			validUpgrades = self.World.WorldActor.TraitOrDefault<ScriptUpgradesCache>();
 		}
 
 		[Desc("Grant an upgrade to this actor.")]
 		public void GrantUpgrade(string upgrade)
 		{
-			um.GrantUpgrade(Self, upgrade, this);
+			if (validUpgrades.Info.Upgrades.Contains(upgrade))
```

Comment:
I feel we ought to retain the failing logic here in the case of a bogus upgrade, instead of failing silently.
If the author does not run a lint pass that is their fault and should be reported to them (via crashes).